drivers/platform/stm32/stm32_spi.c

```diff
@@ -386,12 +430,134 @@ int32_t stm32_spi_write_and_read(struct no_os_spi_desc *desc,
+				   struct no_os_spi_msg* msgs,
+				   uint32_t len)
+{
+#ifdef HAL_DMA_MODULE_ENABLED
+	struct stm32_spi_desc* sdesc = desc->extra;
+	struct no_os_dma_xfer_desc* rx_ch_xfer;
+	struct no_os_dma_xfer_desc* tx_ch_xfer;
+	struct no_os_dma_ch* tx_ch = sdesc->txdma_ch;
+	struct no_os_dma_ch* rx_ch = sdesc->rxdma_ch;
+	struct stm32_dma_channel* sdma_rx = rx_ch->extra;
+	struct stm32_dma_channel* sdma_tx = tx_ch->extra;
+	SPI_TypeDef* SPIx = sdesc->hspi.Instance;
+	int ret;
+	uint8_t i;
+
+	if (!desc || !msgs)
+		return -EINVAL;
+
+	rx_ch_xfer = no_os_calloc(len, sizeof(*rx_ch_xfer));
+	if (!rx_ch_xfer)
+		return -ENOMEM;
+
+	tx_ch_xfer = no_os_calloc(len, sizeof(*tx_ch_xfer));
+	if (!tx_ch_xfer) {
+		goto free_rx_ch_xfer;
+	}
+
+	for (i = 0; i < len; i++) {
+		tx_ch_xfer[i].src = msgs[i].tx_buff;
+		tx_ch_xfer[i].dst = &(SPIx->DR);
+		tx_ch_xfer[i].xfer_type = MEM_TO_DEV;
+		tx_ch_xfer[i].periph = NO_OS_DMA_IRQ;
+		tx_ch_xfer[i].length = msgs[i].bytes_number;
+
+		rx_ch_xfer[i].dst = msgs[i].rx_buff;
+		rx_ch_xfer[i].src = &(SPIx->DR);
+		rx_ch_xfer[i].periph = NO_OS_DMA_IRQ;
+		rx_ch_xfer[i].xfer_type = DEV_TO_MEM;
+		rx_ch_xfer[i].length = msgs[i].bytes_number;
+	}
+
+	rx_ch->id = sdma_rx->hdma;
+	rx_ch->sg_list = NULL;
+
+	ret = no_os_dma_config_xfer(sdesc->dma_desc, rx_ch_xfer, len, rx_ch);
+	if (ret)
+		goto remove_dma;
+
+	tx_ch->id = sdma_tx->hdma;
+	tx_ch->sg_list = NULL;
+
+	ret = no_os_dma_config_xfer(sdesc->dma_desc, tx_ch_xfer, len, tx_ch);
+	if (ret)
+		goto remove_dma;
+
+	ret = no_os_dma_xfer_start(sdesc->dma_desc, tx_ch);
+	if (ret)
+		goto abort_transfer;
+
+	ret = no_os_dma_xfer_start(sdesc->dma_desc, rx_ch);
+	if (ret)
+		goto abort_transfer;
+
+	if (tx_ch)
+		SET_BIT(sdesc->hspi.Instance->CR2, SPI_CR2_TXDMAEN);
+
+	if (rx_ch)
+		SET_BIT(sdesc->hspi.Instance->CR2, SPI_CR2_RXDMAEN);
+
+#ifdef HAL_TIM_MODULE_ENABLED
+	if (sdesc->pwm_desc) {
+		ret = no_os_pwm_enable(sdesc->pwm_desc);
+		if (ret)
+			goto abort_transfer;
+	}
+#endif
+
+	return 0;
+
+abort_transfer:
+	no_os_dma_xfer_abort(sdesc->dma_desc, tx_ch);
+	no_os_dma_xfer_abort(sdesc->dma_desc, rx_ch);
+remove_dma:
+	no_os_dma_remove(sdesc->dma_desc);
+free_tx_ch_xfer:
+	no_os_free(tx_ch_xfer);
+free_rx_ch_xfer:
+	no_os_free(rx_ch_xfer);
+#endif
+
+	return ret;
+}
+
+/**
+ * @brief Configure and start a series of transfers using DMA. Wait for the
+ * 	  completion before returning.
+ * @param desc - The SPI descriptor.
+ * @param msgs - The messages array.
+ * @param len - Number of messages.
+ * @param callback - Function to be invoked once the transfers are done.
+ * @param ctx - User defined parameter for the callback function.
+ * @param is_async - Whether or not the function should wait for the completion.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t stm32_spi_dma_transfer_async(struct no_os_spi_desc* desc,
+				     struct no_os_spi_msg* msgs,
+				     uint32_t len,
+				     void (*callback)(void*),
+				     void* ctx, bool is_async)
```

Comment:
Same as before, this callback is totally ignored and the `stm32_spi_dma_transfer_async` function doesn't take the same parameters as `no_os_spi_transfer_dma_async`, so there is no way this can work.

Reply:
The callback is being registeredwithin the no_os_dma_config_xfer() and enabled via the no_os_dma_xfer_start()

Reply:
@CiprianRegus - what is the difference between the callback that's in the argument here and the one in the no_os_dma_init param?

Reply:
The callback in the `no_os_dma_init_param` struct is meant to be called once a DMA transfer finished (doesn't matter if the transfer was for the SPI peripheral or it was a memory to memory transfer). The callback in the `no_os_spi_transfer_dma_async()` function should be called once the DMA SPI transfer issued by the function call is done.

Reply:
As discussed in the call, I've added a condition to overwrite the desc->sg_handler if the user provides a callback in the function argument of the ``no_os_spi_transfer_dma_async()`` .